Adjacent clusters

bdenckla opened the pull request:
This pull request adds a new CCS (collision candidate set): adjacent clusters.

The existing "marks" CCS, by design, only tests for collisions between marks.

The "faraway" CCS tests for some base-to-mark collisions, but only for marks that are far away, i.e. only for marks that don't belong to the base in question. (Judging from the documentation, I'm not sure whether those faraway base-to-mark tests are even intentional, i.e. judging from the documentation, "faraway" seems to have been intended to test base-to-base collisions only.)

The "cursive" CCS and "area" CCS, by design, only consider adjacent glyphs. Of course this includes some base-to-mark tests, but not all.

So the new adjacent clusters CCS fills what is, at least for my needs, and important void in the options.

The new adjacent clusters CCS truly uses the cluster index from the glyph info rather than using the base and/or mark category. I found this was important for my application, which involves fonts that sometimes insert a thin space in the glyph buffer. This thin space is categorized as a base, but fortunately does not introduce a new cluster.

simoncozens commented:
Oh, that's very nice. It'll need some documentation in `__init__.py` and command-line options in the `__main__.py`; I'm happy to add those, then merge this, unless you have capacity to write them.

bdenckla commented:
Okay, I added the "adjacent clusters" feature to some comments and to `argparse`.

I wasn't sure whether to add it to `argparse` negatively (`--no-x`/`store_false`) or positively (`--x`/`store_true`). I added it negatively. 

The CCS (collision candidate set) features now have quite a bit of overlap, meaning if they are all on, they do a lot of redundant work now. So I'm not sure if adding "adjacent clusters" negatively, i.e. adding `--no-adjacent-clusters`, was the best thing to do.

BTW what is your take on this question I raised?

> I'm not sure whether those faraway base-to-mark tests are even intentional, i.e. judging from the documentation, "faraway" seems to have been intended to test base-to-base collisions only.

simoncozens commented:
The idea behind faraway is to look at interactions between clusters. I may have documented it badly. :-)
